Fix issues with generalized_scatter  and setitem allocated unbacked symbols.

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164948
* __->__ #164341

Three fixes:
1. When doing t[u0] +=1  if u0 is unbacked we could allocate a new unbacked symbol during the the indexing of t[u0] (when we fake trace setitem), namely because meta_select does allocate a new unbacked symbol for the storage offset when we do not know if u0>=0 or u0<0.  but the output size/stride of setitem(), does not depend on that new symbol. it's self consumed in setitem so we shall ignore it. 

2. Also when we trace through generalized_scatter the applications of the views could allocate unbacked symints
but those do not effect final output, we also shall ignore them.

3.Before accessing strides in lowering we shall materialize. 

Address  https://github.com/pytorch/pytorch/issues/114293 and https://github.com/pytorch/pytorch/issues/131911

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

laithsakka committed:
Fix issues with selec_scatter on unbacked

[ghstack-poisoned]

laithsakka commented on torch/_inductor/lowering.py:
```diff
@@ -3141,8 +3142,14 @@ def select_scatter(x, src, dim: int, index: int):
+    elif V.graph.sizevars.guard_or_false(sympy.Ge(index, 0)):
+        pass
+    else:
+        # unabacked index
```
unabacked->unbacked.

laithsakka commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1096,7 +1097,18 @@ def method___setitem__(self, key, value):
 
             # Not sure if __setitem__ can ever save activations, disabling just in case
-            with torch._dynamo.utils._disable_saved_tensors_hooks_during_tracing():
+            # with torch._dynamo.utils._disable_saved_tensors_hooks_during_tracing():
```
remove this line

laithsakka committed:
Update on "Fix issues with select_scatter on unbacked and item allocated unbacked symbols."



Address https://github.com/pytorch/pytorch/issues/162110 and https://github.com/pytorch/pytorch/issues/114293

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "Fix issues with select_scatter on unbacked and item allocated unbacked symbols."


Three fixes:
1. When doing t[x] +=1  if x is unbacked we could allocate a new unbacked symbol during the the indexing of t[x]  but 
the output size/stride does not depend on it in this case. it's self consumed during meta tracing so we shall ignore it. 

2. Also when we trace through generalized_scatter the applications of the views could allocate unbacked symints
but those do not effect final output, we also shall ignore them.

3.Before accessing strides in lowering we shall materialize. 

Address https://github.com/pytorch/pytorch/issues/162110 and https://github.com/pytorch/pytorch/issues/114293

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "Fix issues with select_scatter  and update_item allocated unbacked symbols."


Three fixes:
1. When doing t[x] +=1  if x is unbacked we could allocate a new unbacked symbol during the the indexing of t[x]  but 
the output size/stride does not depend on it in this case. it's self consumed during meta tracing so we shall ignore it. 

2. Also when we trace through generalized_scatter the applications of the views could allocate unbacked symints
but those do not effect final output, we also shall ignore them.

3.Before accessing strides in lowering we shall materialize. 

Address  https://github.com/pytorch/pytorch/issues/114293 and https://github.com/pytorch/pytorch/issues/131911

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1096,7 +1097,18 @@ def method___setitem__(self, key, value):
+
+            with (
+                torch._dynamo.utils._disable_saved_tensors_hooks_during_tracing(),
+                tx.fake_mode.shape_env.ignore_fresh_unbacked_symbols()
```
ok let me explain this in details
When we trace   wrk[idx] += 1 this is what happen : 
```

TRACE LOAD_FAST wrk [RangeIteratorVariable()]
TRACE LOAD_FAST idx [RangeIteratorVariable(), TensorVariable()]
TRACE DUP_TOP_TWO None [RangeIteratorVariable(), TensorVariable(), SymNodeVariable()]
TRACE BINARY_SUBSCR None [RangeIteratorVariable(), TensorVariable(), SymNodeVariable(), TensorVariable(), SymNodeVariable()]      create u1 for the select operation since input to is unbacked and use it in the storage offset of  output of TRACE BINARY_SUBSCR  TRACE LOAD_CONST 1 [RangeIteratorVariable(), TensorVariable(), SymNodeVariable(), TensorVariable()] TRACE INPLACE_ADD None [RangeIteratorVariable(), TensorVariable(), SymNodeVariable(), TensorVariable(), ConstantVariable(int: 1)] .. TRACE STORE_SUBSCR None [RangeIteratorVariable(), TensorVariable(), TensorVariable(), SymNodeVariable()]      
```
 reach this line of code, during  get_fake_value(proxy.node, tx, allow_non_graph_fake=False)      proxy.node is set_item when we call we call select to access the pos to write but the out of the set_item does not care about u2 allocated in  that select.

laithsakka committed:
Update on "Fix issues with select_scatter  and update_item allocated unbacked symbols."


Three fixes:
1. When doing t[x] +=1  if x is unbacked we could allocate a new unbacked symbol during the the indexing of t[x]  but 
the output size/stride does not depend on it in this case. it's self consumed during meta tracing so we shall ignore it. 

2. Also when we trace through generalized_scatter the applications of the views could allocate unbacked symints
but those do not effect final output, we also shall ignore them.

3.Before accessing strides in lowering we shall materialize. 

Address  https://github.com/pytorch/pytorch/issues/114293 and https://github.com/pytorch/pytorch/issues/131911

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

ezyang commented:
I don't understand the pr description for (1). What does it mean that you get a new unbacked symbol if x is unbacked?

laithsakka committed:
Update on "Fix issues with generalized_scatter  and setitem allocated unbacked symbols."


Three fixes:
1. When doing t[u0] +=1  if x is unbacked we could allocate a new unbacked symbol during the the indexing of t[u0] (when we fake trace setitem), namely because meta_select does allocate a new unbacked symbol for the storage offset when we do not know if u0>=0 or u0<0.  but the output size/stride of setitem(), does not depend on that new symbol. it's self consumed in setitem so we shall ignore it. 

2. Also when we trace through generalized_scatter the applications of the views could allocate unbacked symints
but those do not effect final output, we also shall ignore them.

3.Before accessing strides in lowering we shall materialize. 

Address  https://github.com/pytorch/pytorch/issues/114293 and https://github.com/pytorch/pytorch/issues/131911

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

ezyang reviewed (changes requested):
vc me if you don't agree

bobrenjc93 commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1096,7 +1097,19 @@ def method___setitem__(self, key, value):
-            with torch._dynamo.utils._disable_saved_tensors_hooks_during_tracing():
+
+            # Ignore fresh unbacked symbols that could arise from the internal indexing (selection),
+            # that happen in code like t[idx] += 1 when idx is unabacked. Namely the selection
```
s/unabacked/unbacked

bobrenjc93 commented on torch/_inductor/lowering.py:
```diff
@@ -1946,6 +1946,7 @@ def select(x, dim, idx):
             # Additionally, we want to avoid accidental unbacked unsqueeze semantics. To resolve this,
             # we use as_strided instead.
             # Removing this branch will cause test_unbacked_select_index_with_check to fail.
+            x.realize()
```
add a comment?

bobrenjc93 commented on torch/_inductor/lowering.py:
```diff
@@ -1968,7 +1969,7 @@ def select(x, dim, idx):
     assert len(unbacked_bindings) == 1, unbacked_bindings
     unbacked_offset_sym, _ = next(iter(unbacked_bindings.items()))
-
+    x.realize()
```
ditto

laithsakka committed:
Update on "Fix issues with generalized_scatter  and setitem allocated unbacked symbols."


Three fixes:
1. When doing t[u0] +=1  if u0 is unbacked we could allocate a new unbacked symbol during the the indexing of t[u0] (when we fake trace setitem), namely because meta_select does allocate a new unbacked symbol for the storage offset when we do not know if u0>=0 or u0<0.  but the output size/stride of setitem(), does not depend on that new symbol. it's self consumed in setitem so we shall ignore it. 

2. Also when we trace through generalized_scatter the applications of the views could allocate unbacked symints
but those do not effect final output, we also shall ignore them.

3.Before accessing strides in lowering we shall materialize. 

Address  https://github.com/pytorch/pytorch/issues/114293 and https://github.com/pytorch/pytorch/issues/131911

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka commented:
address comments and rebase

laithsakka commented:
@pytorchmergebot merge
